inginious/agent/docker_agent/__init__.py

```diff
@@ -505,6 +517,46 @@ async def create_student_container(self, parent_info, socket_id, environment_nam
         except:
             self._logger.exception("Exception in create_student_container")
 
+    async def start_ssh(self, read_stream, info):
+        """ Wait for ssh information from student_container and send ssh info to frontend """
+        buffer = bytearray()
+        try:
+            while not read_stream.at_eof():  # CHECK HERE
+                msg_header = await read_stream.readexactly(8)
+                outtype, length = struct.unpack_from('>BxxxL',
+                                                     msg_header)  # format imposed by docker in the attach endpoint
+                if length != 0:
+                    content = await read_stream.readexactly(length)
+                    if outtype == 1:  # stdout
+                        buffer += content
+
+                    if outtype == 2:  # stderr
+                        self._logger.debug("Received stderr from containers:\n%s", content)
+                    # 4 first bytes are the length of the message. If we have a complete message...
+                    while len(buffer) > 4 and len(buffer) >= 4 + struct.unpack('!I', buffer[0:4])[0]:
+                        msg_encoded = buffer[4:4 + struct.unpack('!I', buffer[0:4])[0]]  # ... get it
+                        buffer = buffer[4 + struct.unpack('!I', buffer[0:4])[0]:]  # ... withdraw it from the buffer
```

Comment:
``struct.unpack('!I', buffer[0:4])[0]``seems to be computed multiple times here without ``buffer`` being modified in the meantime. It looks like a length ? Maybe it would be more readable to have it in an intermediate variable.

Reply:
Using intermediate "length" variable.
Anyway, this hard to read part is now included in a small function making it more explicit.